Correctly update a subscription status upon losing a dispute

wjrosa opened the pull request:
Fixes 4465-gh-woocommerce/woocommerce-subscriptions

## Description

<!--
Write a brief summary about this PR. 
- Why is this change needed? 
- What does this change do? 
- Were there other solutions you considered? 
- Why did you choose to pursue this solution? 
- Describe any trade-offs you might have had to make.
-->

This PR correctly changes a subscription status when the related order loses a dispute. This is done by checking the current status of the subscription before updating it. If the status is `pending-cancel` we just update it to `cancelled`. 
<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos as appropriate.
-->

## How to test this PR

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
Use the testing instructions from the linked issue as a starting point.
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Code review should be enough. Check if the tests are still passing. I have included specific unit tests for this change.

## Product impact
<!-- What products will this PR ship in? -->

- [x] Added changelog entry (or does not apply)
- [x] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [x] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref
- [ ] <!-- 🚨 Deprecations 🚨 --> Added deprecated functions, hooks or classes to the [spreadsheet](https://docs.google.com/spreadsheets/d/1xw9xszcPMnWsp4C8OKZMsLzZob7tOmWT7qMqmEIq314/edit#gid=0)

wjrosa committed:
Correctly update a subscription status upon losing a dispute

wjrosa committed:
Checking for existing meta instead of note

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Adding changelog entry

mattallan reviewed (changes requested):
@wjrosa thanks for opening this PR!

The `_dispute_closed_status` order meta is only set and used by the Stripe gateway and so we shouldn't be writing changes to WooCommerce Subscriptions to support this specific meta.

WooCommerce Subscriptions, needs to be as gateway agnostic as possible, otherwise we'll have to write 50-60+ gateway support into subscriptions 😅 

---

Looking at the original GH issue, my understanding is that the problem has to do with our extension trying to mark the subscription as `on-hold` when calling `$subscription->payment_failed()` when the subscription already has `pending-cancel`.

To address this, I think our approach should be to update our `failed_subscription_sign_ups_for_order()` to check the status of the subscription before calling `$subscription->payment_failed()` and if the status is already `pending-cancel`, mark it as `cancelled`.

Or maybe we should update the default status in `$subscription->payment_failed()` so that it uses cancelled if the subscription is pending cancelled.

We will also want to look at updating the `payment_failed()` function because currently it adds an order note saying the subscription was cancelled due to maximum failed payments reached, which isn't relevant to this and many other cases 🤔 

![image](https://github.com/user-attachments/assets/fa7d93ef-55b3-4ff0-9260-c68ea0774241)

I think this order note should only be used if the `woocommerce_subscription_max_failed_payments_exceeded` filter hook was used.

wjrosa committed:
Improving implementation after review

wjrosa committed:
Simplifying filter logic and fixing the inclusion of failed attempts note

wjrosa commented:
> @wjrosa thanks for opening this PR!
> 
> The `_dispute_closed_status` order meta is only set and used by the Stripe gateway and so we shouldn't be writing changes to WooCommerce Subscriptions to support this specific meta.
> 
> WooCommerce Subscriptions, needs to be as gateway agnostic as possible, otherwise we'll have to write 50-60+ gateway support into subscriptions 😅
> 
> Looking at the original GH issue, my understanding is that the problem has to do with our extension trying to mark the subscription as `on-hold` when calling `$subscription->payment_failed()` when the subscription already has `pending-cancel`.
> 
> To address this, I think our approach should be to update our `failed_subscription_sign_ups_for_order()` to check the status of the subscription before calling `$subscription->payment_failed()` and if the status is already `pending-cancel`, mark it as `cancelled`.
> 
> Or maybe we should update the default status in `$subscription->payment_failed()` so that it uses cancelled if the subscription is pending cancelled.
> 
> We will also want to look at updating the `payment_failed()` function because currently it adds an order note saying the subscription was cancelled due to maximum failed payments reached, which isn't relevant to this and many other cases 🤔
> 
> ![image](https://private-user-images.githubusercontent.com/2275145/369736066-fa7d93ef-55b3-4ff0-9260-c68ea0774241.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcwOTEzMjQsIm5iZiI6MTcyNzA5MTAyNCwicGF0aCI6Ii8yMjc1MTQ1LzM2OTczNjA2Ni1mYTdkOTNlZi01NWIzLTRmZjAtOTI2MC1jNjhlYTA3NzQyNDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjNUMTEzMDI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2FkZGJhMWI2Zjc0YmM5YmJmOTliOTBjYTRmOWU2OWNkYTM3NmY5MTgzMjhhNzZmNjAwNDRmY2JkN2VhYzQ1ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.pYlWXgCHrD1cySzSGg2upCwQ5IeOrGtYM2Fx7y4mlJE)
> 
> I think this order note should only be used if the `woocommerce_subscription_max_failed_payments_exceeded` filter hook was used.

My initial thought was to use a very neutral name for the meta and document this new meta somewhere for other gateways to use it. But yeah, your suggestion is much simpler and smarter! Thank you so much! I have updated the code here and I will close the Stripe extension PR.

Also, I have simplified the filter code you mentioned and changed it to add the note only when the filter is executed in https://github.com/Automattic/woocommerce-subscriptions-core/pull/687/commits/70276dbaee5ffda3593b27716ca499d6af7d5fd5

wjrosa committed:
Update changelog entry

mattallan commented on includes/class-wc-subscription.php:
```diff
@@ -1913,12 +1913,13 @@ public function payment_failed( $new_status = 'on-hold' ) {
-			$this->update_status( $new_status );
+		$status_note = '';
+		if ( apply_filters( 'woocommerce_subscription_max_failed_payments_exceeded', false, $this ) ) {
+			$status_note = __( 'Subscription Cancelled: maximum number of failed payments reached.', 'woocommerce-subscriptions' );
```
We need to also set `$new_status = 'cancelled';` when this filter is used to preserve backwards compatibility.

wjrosa replied:
Gotcha! I have updated the whole implementation again to make this small change easier to read in https://github.com/Automattic/woocommerce-subscriptions-core/pull/687/commits/b7a92f35657d25a5480fdf8e4f15c753506ffd22

mattallan reviewed (changes requested):
Thanks @wjrosa for the later changes.

I noticed a small backwards compatibility issue for any gateway using the `woocommerce_subscription_max_failed_payments_exceeded` filter.

wjrosa committed:
Fix backwards compatibility

mattallan reviewed (changes requested):
Hey @wjrosa, so sorry to be a pain but this latest commit doubles the size of the PR to include PHPCS adjustments that I think should be handled in a separate PR so that we can discuss and review those changes on their own 😅

It's generally fine to address a small whitespace/PHPCS tweaks in the function you're updating, however, this PR is now updating all instances of `current_time( 'timestamp', true )` in the main subscriptions class which makes it difficult to review the changes  in this PR and understand which code is relevant to the fix.

Also if there's any reason we need to revisit these `current_time()` changes if it's in a separate PR, we can easily find/review the changes and revert it without that impacting other fixes :)

With that in mind, could you please remove the fixes to PHPCS violation fixes and create an issue/PR for the PHPCS changes? i.e. create an issue to replace all instances of `current_time( 'timestamp', true )` with `time()`?

Sorry and thanks 😄

mattallan commented on includes/class-wc-subscription.php:
```diff
@@ -1913,12 +1913,18 @@ public function payment_failed( $new_status = 'on-hold' ) {
-				$this->update_status( 'cancelled', __( 'Subscription Cancelled: maximum number of failed payments reached.', 'woocommerce-subscriptions' ) );
+		// This also forces the new status to be 'cancelled' if the filter is applied or the subscription is pending-cancel
+		$status_note                                 = '';
+		$max_failed_payments_exceeded_filter_applied = apply_filters( 'woocommerce_subscription_max_failed_payments_exceeded', false, $this );
```
Small nit-pick: the large amount of whitespace here to align `=` looks a bit funky 🥲 What are your thoughts on shortening this variable name to something like: `$max_failed_payments`?

I was thinking we could also remove the nested if statement needed to set the status note by doing something like:

```
$max_failed_payments = apply_filters( 'woocommerce_subscription_max_failed_payments_exceeded', false, $this );
$status_note         = $max_failed_payments ? __( 'Subscription Cancelled: maximum number of failed payments reached.', 'woocommerce-subscriptions' ) : '';

if ( $max_failed_payments || $this->has_status( 'pending-cancel' ) ) {
	$new_status = 'cancelled';
}
```

Not sure if this is more readable or worse though. Happy to run with what you think is best :)

wjrosa replied:
Yeah, I agree. Usually, it is best practice to name variables according to exactly what they are in PHP (citing PHP here because in Go, for example, it would be OK to name this `mfp`), even if it gets long. So, personally, my issue is more with the attribution alignment rule than with the variable name itself 😅 , but I understand this is a global rule. Change applied ✅ 

I loved your suggestion! I think it is simpler to understand 👍 . Also applied

wjrosa committed:
Reverting auto lint fixes

wjrosa committed:
Improving code readability

wjrosa commented:
> Hey @wjrosa, so sorry to be a pain but this latest commit doubles the size of the PR to include PHPCS adjustments that I think should be handled in a separate PR so that we can discuss and review those changes on their own 😅
> 
> It's generally fine to address a small whitespace/PHPCS tweaks in the function you're updating, however, this PR is now updating all instances of `current_time( 'timestamp', true )` in the main subscriptions class which makes it difficult to review the changes in this PR and understand which code is relevant to the fix.
> 
> Also if there's any reason we need to revisit these `current_time()` changes if it's in a separate PR, we can easily find/review the changes and revert it without that impacting other fixes :)
> 
> With that in mind, could you please remove the fixes to PHPCS violation fixes and create an issue/PR for the PHPCS changes? i.e. create an issue to replace all instances of `current_time( 'timestamp', true )` with `time()`?
> 
> Sorry and thanks 😄

There's no need to apologize! Thanks for disclosing this kind of issue. I don't know if it's because I am used to automatically fixing lint issues (using my editor) on the files I work on or if it was the git commit hook. Still, it was not fully intentional. I reverted those changes now.

I don't even know why the `current_time` calls were changed. Do you have more info on this? Then I can work on those on another PR.

wjrosa committed:
Merge branch 'trunk' into fix/correctly-update-subs-status-upon-losing-dispute

mattallan reviewed (approved):
Thanks @wjrosa !

I've tested these changes by:
1. Purchase a subscription
2. Mark the new subscription as pending-cancellation
3. Go to the parent order and mark it as failed
4. On `trunk` the subscription remains pending cancellation
5. On this branch the subscription transitions to cancelled 👍
